openedx/openedx-platform
# APIs for XBlock static asset files in Blockstore

This PR introduces some backend python+REST APIs for storing static asset files along with an XBlock in a content library. It also updates the runtime to be able to load such static asset files.

Example use cases:
* Store an image file with an HTML block and then use the image inline in the HTML block.
* Store a PDF file with an HTML block and provide a link in the HTML for the learner to download the PDF.
* Store .srt files or even video .mp4 files that belong to a video XBlock.

This is currently only working for individual XBlocks in blockstore-based content libraries, but something like this is meant to eventually replace the "Files & Uploads" tool that exists in Studio today, which tends to get very cluttered and disorganized (because it's difficult to tell what files are in use by what XBlocks).

Note that it's fine to upload the same file to multiple XBlocks, because Blockstore will automatically de-duplicate the uploads. In the future we'll want a UI to allow directly "copying" a file from another XBlock instead of requiring the user to upload it multiple times, in cases where the user wants to use a single asset in many different blocks.

I was originally intending for this PR to also introduce support for the `video` XBlock to load transcript files from Blockstore but have held off on that pending questions about using VAL for that instead.

### Screenshot:
This is from [Ramshackle](https://github.com/open-craft/ramshackle/), the developer UI for testing content libraries. It has already been updated to work with these new APIs.  

![Screen Shot 2019-10-23 at 4 48 47 PM](https://user-images.githubusercontent.com/945577/67432852-07304380-f5b5-11e9-94d2-a6caf40c43ff.png)


### Test instructions:

HTML block:
* Update Ramshackle and access it at http://localhost:18010/ramshackle/
* Add/edit an HTML block in Ramshackle.
* From the "Assets" tab, drag and drop an image file.
* From the "Source" tab, add some HTML to reference the image file like `<img src="/static/the_filename.png" alt="An image I just added" />`.
* Save and go to the "View" tab. You should now see the image. Go back to the library's list of blocks and publish the library changes. No view this HTML block in the "Learn" tab, to confirm the image also works when published and when using the LMS APIs (different than when in draft mode and using the Studio APIs).

One can also test this with Amazon S3 by setting the following in Blockstore's `settings/private.py`:
```
DEFAULT_FILE_STORAGE = "storages.backends.s3boto3.S3Boto3Storage"
AWS_ACCESS_KEY_ID = "changeme"
AWS_SECRET_ACCESS_KEY = "changeme"
AWS_STORAGE_BUCKET_NAME = "changeme"
```

### Integration Test Instructions

Because the tests require a running blockstore instance, they won't run by default. To run them on your devstack:

From the `blockstore` directory on your host, run `make testserver` to get a blockstore test server running. Then From `make studio-shell`, run this test command:

```
EDXAPP_RUN_BLOCKSTORE_TESTS=1 python -Wd -m pytest --ds=cms.envs.test openedx/core/lib/blockstore_api/ openedx/core/djangolib/tests/test_blockstore_cache.py openedx/core/djangoapps/content_libraries/tests/ common/lib/xmodule/xmodule/tests/test_unit_block.py
```

Then repeat the command from `make lms-shell`, with `cms.envs.test` changed to `lms.envs.test`.
